cmd/output/progress/progress.go

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2014-2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package progress
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+// Meter is an interface to show progress to the user
+type Meter interface {
+	io.Writer
+
+	// Start progress with max "total" steps.
+	Start(label string, total float64)
+
+	// Set progress to the "current" step.
+	Set(current float64)
+
+	// SetTotal sets "total" steps needed.
+	SetTotal(total float64)
+
+	// Finished the progress display
+	Finished()
+
+	// Spin indicates indefinite activity by showing a spinner.
+	Spin(msg string)
+
+	// Notify the user of miscellaneous events
+	Notify(string)
+}
+
+// MakeProgressBar creates an appropriate progress.Meter for the environ in
+// which it is called:
+//
+// * if no terminal is attached, or we think we're running a test, a
+//   minimalistic QuietMeter is returned.
+// * otherwise, an ANSIMeter is returned.
+//
+func MakeProgressBar(stdout io.Writer) Meter {
+	if terminal.IsTerminal(int(os.Stdin.Fd())) {
+		return &ANSIMeter{
+			terminal:    term{},
+			stdout:      stdout,
+			escapeChars: DefaultEscapeChars(),
```

Comment:
Are we sure these are valid under windows? They seem fairly trivial escape sequences but I don't have access to a windows instance to test.

Reply:
Looks like they're not -- or at least something in here is causing these tests to fail on Windows. For example: https://jenkins.juju.canonical.com/job/github-unit-tests-win2012/6726/testReport/github/com_juju_juju_cmd_output_progress/Test/